### ISSUE TITLE
Essential Cross compiling iOS Urgent need to answer

### ISSUE THREAD

**Issue by iOSzhangkai** [MTG/essentia#620]
I cross compile iOS practices:
1.

> brew, install, yasm, cmake, WGet, packaging/build_3rdparty_static_debian.sh

2.

> Building, lightweight, Essentia, with, reduced, dependencies:./waf, configure, --lightweight=, --fft=ACCELERATE

3.iOS:

> ./waf, configure, --cross-compile-ios, --lightweight=, --fft=ACCELERATE, --build-static

Is that correct? I have libessentia.a. in the /essentia-master/build/src directory, but how does this libessentia.a static package work in iOS's Xcode project?

**Comment by carthach** [comment]
I answered this in your other message.

**Comment by iOSzhangkai** [comment]
Excuse me. You gave me this project. I saw it on someone else's problem yesterday, but after the download of this project, there was no file in the iOS directory;
Cross compilation of iOS official documents involving less, but there is no system compiler tutorial, do you have any blog or document you want to study, I hope you can let me learn, thank you

**Comment by iOSzhangkai** [comment]
I follow the document on the iOS cross compiler command, libessentia.a file in the /build/src directory, the libessentia.a package is generally with a static.H file, is my error allocation process, not playing.H files, or what reason, I hope you can help me to answer, thank you 
[PS: I'm not too good the English, if there is no clear expression, please forgive]

**Comment by carthach** [comment]
Sorry I can't really understand your problem and I am not an iOS developer so I can't really help any more.

You will need to search Google for information on using C++ libraries with iOS.

For example here's a tutorial I found:
https://www.sitepoint.com/using-c-and-c-in-an-ios-app-with-objective-c/

Good luck!

**Comment by willrevl** [comment]
So I think I have hit the same issue as him. I've got the essentia complied with the flags outlined in the guide. But where are the header files that we need to use this library? When using a static library in Xcode, there are some header files which we can use to interface with the library from objective-c. I saw somewhere else on that these are located in `/user/local/include/essentia` however I do not see the `essentia` directory here. Is there an additional flag needed to generate these?

**Comment by carthach** [comment]
Well remember to set the prefix flag --prefix= to wherever you want.

After you compile don't forget to run ./waf install. This will install into the location and will install headers into

<LOCATION>/include

and the library into:

<LOCATION>/lib

**Comment by willrevl** [comment]
@carthach thanks so much for the prompt response. I found the headers! 

I'm really close, my problem now is that I can't get Xcode to see the c++ stl. argh!

**Comment by willrevl** [comment]
If anyone is feeling super helpful, here is a sample project with essential and the errors I'm getting: https://www.dropbox.com/s/h3dw26a12w1fpne/AudioAnalysis.zip?dl=0

**Comment by willrevl** [comment]
^ my issue there was that I needed to include essential from a .mm file so that Xcode would recognize it as objective-c++

**Comment by carthach** [comment]
Might need to include libstdc++ as well, I'm not sure. I'm not an iOS programmer I'm afraid! I just compiled it for use with JUCE which allows you to program directly for iOS using C++.

**Comment by willrevl** [comment]
Xcode already knows about libstdc++, it just needs to interpret the file you are using as objective-c++ to know it needs to use it.

**Comment by willrevl** [comment]
So I just tried to sort of copy the example from here: https://github.com/MTG/essentia/blob/master/src/examples/onset_detector.cpp

but I got this error message: `libc++abi.dylib: terminating with uncaught exception of type essentia::EssentiaException: Identifier 'MonoLoader' not found in registry...`

Are the examples working with an older version of essential or something? Or might I be doing something horribly wrong?

**Comment by dbogdanov** [comment]
The MonoLoader algorithm is missing if you compiled essentia without ffmpeg
dependency. See the output of wad configure or check build/config.log for
more details.

**Comment by willrevl** [comment]
Yes, seems this is the default for using on iOS, is there an alternative to the MonoLoader algorithm for getting audio data in?

**Comment by willrevl** [comment]
I suppose just read it with audio units then convert to a vector<Real> and that should work.

**Comment by willrevl** [comment]
Another question, my project already depends on ffmpeg, is it possible to do the iOS compile, but include the ffmpeg dependency?

**Comment by dbogdanov** [comment]
You can specify dependencies to include using the --lightweight flag. For
example, --lightweight=libav will include ffmpeg (build script will search
for it using pkg-config)

**Comment by willrevl** [comment]
OK, so i've tried building with this flag. However now I am getting this: 
```
./waf configure --cross-compile-ios --lightweight=libav --fft=ACCELERATE --build-static
```
```
...
- FFmpeg / libav detected!
- libsamplerate seems to be missing.
  The following algorithms will be ignored: ['Resample', 'MonoLoader', 'EqloudLoader', 'EasyLoader']
```

**Comment by willrevl** [comment]
When I check for libsamplerate:
[Revl] essentia$ brew install libsamplerate
Warning: libsamplerate 0.1.9 is already installed

**Comment by willrevl** [comment]
So I added libsamplerate and the build appeared to include the 'MonoLoader' algorithm. However, now i'm getting:

```
Undefined symbols for architecture arm64:
  "_src_reset", referenced from:
      essentia::streaming::Resample::reset() in libessentia.a(resample.cpp.1.o)
  "_src_simple", referenced from:
      essentia::standard::Resample::compute() in libessentia.a(resample.cpp.1.o)
  "_src_process", referenced from:
      essentia::streaming::Resample::process() in libessentia.a(resample.cpp.1.o)
  "_src_delete", referenced from:
      essentia::streaming::Resample::~Resample() in libessentia.a(resample.cpp.1.o)
      essentia::streaming::Resample::configure() in libessentia.a(resample.cpp.1.o)
  "_src_new", referenced from:
      essentia::streaming::Resample::configure() in libessentia.a(resample.cpp.1.o)
  "_src_strerror", referenced from:
      essentia::standard::Resample::compute() in libessentia.a(resample.cpp.1.o)
      essentia::streaming::Resample::process() in libessentia.a(resample.cpp.1.o)
      essentia::streaming::Resample::reset() in libessentia.a(resample.cpp.1.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by willrevl** [comment]
I'm assuming I need to compile libsamplerate, but I am stuck at getting this compiled for arm64

**Comment by willrevl** [comment]
for the next guy, finally got this working steps where:

1. Compile essentia
`./waf configure --cross-compile-ios —lightweight=libav,libsamplerate --fft=ACCELERATE --build-static`
2. Compile FFMPEG with libavsamplerate
3. clone https://github.com/erikd/libsamplerate, drag files from source into a new Xcode project of type static library / framework and compile. Use the resulting .a file.

This will allow the MonoLoader algorithm and so examples can be copied more easily:

https://github.com/MTG/essentia/blob/master/src/examples/

**Comment by carthach** [comment]
Nice work Will!

**Comment by nbranzburg** [comment]
I hope I am posting this in the right place, but it looks like this thread is relevant to what I am struggling with. I am currently working on an iOS app in Swift. It's in it's early stages, and I am very interested in using the Essentia library as part of the app. I have been trying to understand how I can cross compile into iOS and integrate the library into my Swift project.  I am new to Essentia, iOS, Swift, and haven't worked with C++ in years, so please excuse me if I seem a little slow on the uptake here.

I successfully ran the `./waf configure --cross-compile-ios --lightweight= --fft=ACCELERATE --build-static` command. It produced a folder called "build" with the following contents:

- .lock-waf_darwin_build
- .wafpickle-darwin-34015984-98
- c4che (folder)
- config.log (folder)
- essentia.pc
- src (folder)

Inside the src folder I find the following:

- 3rdparty (folder)
- algorithms (folder)
- essentia (folder)
- libessentia.a

It's my understanding that libessentia.a is the compiled C++ library which includes all the object files in the other folders. After reading [this article](http://www.swiftprogrammer.info/swift_call_cpp.html) it seems that I would need to include this file in the **Build Phases** section of my project's configuration file. It also seems that I would need the header (.h) files for Essentia. Now, I ran the `./waf install` command, and I found some header files under **/usr/local/include/essentia**, however it doesn't look like they are all there to me: I only see matching header files for the contents of the **essentia** folder, but none for the other three folders under the **src** folder.

In any case, I added them to my project and linked the library file in the configuration, but I get a build error that says:

> ld: library not found for -lessentia
> clang: error: linker command failed with exit code 1 (use -v to see invocation)

The method in the article also seems really labor intensive, as it involves creating a wrapper for every method/function you want to use.

I'm really stumped with this. Any help or advice would be greatly appreciated!

**Comment by pasotee** [comment]
Hi @willrevl and @nbranzburg I am the next guys and I am very thankful for your insights on this one, looks like you saved me a lot of time.

Still up to date and working.

Have a great day, gentlemen.

**Comment by rahul140490** [comment]
Hi @willrevl , the steps you've mentioned - 
```
1. Compile essentia ./waf configure --cross-compile-ios —lightweight=libav,libsamplerate --fft=ACCELERATE --build-static
2. Compile FFMPEG with libavresample
3. clone https://github.com/erikd/libsamplerate, drag files from source into a new Xcode project of type static library / framework and compile. Use the resulting .a file.
```

The 2nd step isn't clear to me, how should I compile FFMPEG with libavresample? 
And in 3rd step, should I also include essentia.a file in Xcode project for generating a static library?

**Comment by willswinson** [comment]
@nckbrzg Hey there, did you ever figure out the issue with the missing headers? I am having the same issue. I am able to build my project and create obj-c wrappers around the existing headers fine, but there are missing headers for the algorithms I need and other things. I've confirmed that I am building them with my waf config as well because I explicitly tell it to build them with --include-algos, so they should be there.
